### PR TITLE
feature - new line follower

### PR DESCRIPTION
**Pull request by RobertLucian**
1. Keep everything backwards-compatible while adding support for the new line follower too.
1. Use SW I2C everywhere.
1. Change documentation accordingly.
1. Add project to let you tune your own PID controller for the line follower. It also comes with predefined Kp/Ki/Kd parameters for each line follower.

Requires https://github.com/DexterInd/DI_Sensors/pull/60.

### CODE REVIEW

File: Software/Python/easygopigo3.py [hunk 1]
```diff
@@ -964,10 +964,10 @@ def init_line_follower(self, port = "I2C"):
         """
         Initialises a :py:class:`~easysensors.LineFollower` object and then returns it.
 
-        :param str port = "I2C": The only option for this parameter is ``"I2C"``. The default value for this parameter is already set to ``"I2C"``.
+        :param str port = "I2C": The only option for this parameter for the `Old Line Follower`_ is ``"I2C"`` and for the `Line Follower`_ is also ``"AD1"``/``"AD2"``. The default value for this parameter is already set to ``"I2C"``.
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: for the  "red line follower"

Comment (CleoQc) [reply Software/Python/easygopigo3.py]: and the black line follower

File: Software/Python/easysensors.py [hunk 2]
```diff
@@ -29,14 +29,14 @@ def debug(in_str):
...
-    from line_follower import line_sensor
+    from di_sensors.easy_line_follower import EasyLineFollower
 
     # is_line_follower_accessible not really used, just in case
     is_line_follower_accessible = True
 except:
     try:
-        sys.path.insert(0, '/home/pi/GoPiGo/Software/Python/line_follower')
-        import line_sensor
+        sys.path.insert(0, '/home/pi/DI_Sensors/Python/di_sensors')
```
Comment (CleoQc) [inline Software/Python/easysensors.py]: Do we need to insert this path? Isn't it system wide?

File: Software/Python/easysensors.py [hunk 3]
```diff
@@ -1727,10 +1727,12 @@ def read(self):
 
 class LineFollower(Sensor):
     """
-    Class for interacting with the `Line Follower`_ sensor.
+    Class for interacting with the `Line Follower`_ or the `Old Line Follower`_ sensor.
```
Comment (CleoQc) [inline Software/Python/easysensors.py]: black and red

File: Software/Python/easysensors.py [hunk 4]
```diff
@@ -1727,10 +1727,12 @@ def read(self):
...
 
     With this sensor, you can make your robot follow a black line on a white background.
-    The `Line Follower`_ sensor has 5 IR sensors.
+    The `Line Follower`_ sensor has 6 IR sensors and the `Old Line Follower`_ has 5. This class can
```
Comment (CleoQc) [inline Software/Python/easysensors.py]: three guesses as to what my comment here is supposed to be.. 😃

File: Software/Python/easysensors.py [hunk 5]
```diff
@@ -1744,122 +1746,99 @@ class LineFollower(Sensor):
...
         :param easygopigo3.EasyGoPiGo3 gpg = None: The :py:class:`~easygopigo3.EasyGoPiGo3` object that we need for instantiating this object.
         :param bool use_mutex = False: When using multiple threads/processes that access the same resource/device, mutexes should be enabled.
-        :raises ImportError: If the :py:mod:`line_follower` module couldn't be found.
+        :raises ImportError: If the ``di_sensors`` library couldn't be found.
         :raises TypeError: If the ``gpg`` parameter is not a :py:class:`~easygopigo3.EasyGoPiGo3` object.
         :raises IOError: If the line follower is not responding.
 
-        The only value the ``port`` parameter can take is ``"I2C"``.
-        The I2C ports' location on the `GoPiGo3`_ robot can be seen in the following graphical representation: :ref:`hardware-ports-section`.
+        The only value the ``port`` parameter can take for the `Old Line Follower`_ is the ``"I2C"`` and for the new `Line Follower`_ it can also be ``"AD1"`` and ``"AD2"``.
```
Comment (CleoQc) [inline Software/Python/easysensors.py]: here too

File: Software/Python/easysensors.py [hunk 6]
```diff
@@ -1744,122 +1746,99 @@ class LineFollower(Sensor):
...
         try:
             self.set_descriptor("Line Follower")
             Sensor.__init__(self, port, "INPUT", gpg, use_mutex)
-            if line_sensor.read_sensor() == [-1, -1, -1, -1, -1]:
-                raise IOError("Line Follower not responding")
-            self.white_line = line_sensor.get_white_line()
-            self.black_line = line_sensor.get_black_line()
-            self._calculate_threshold()
+            if port == "AD1" or port == "AD2":
+                self._lf = EasyLineFollower(port, module_id = 2)
```
Comment (CleoQc) [inline Software/Python/easysensors.py]: what's this module_id?   Can't we make it transparent to the user? Like I did for distance sensor?

File: Software/Python/easysensors.py [hunk 7]
```diff
@@ -1744,122 +1746,99 @@ class LineFollower(Sensor):
...
             raise
 
-    def _calculate_threshold(self):
-        self.range_sensor = list(map(operator.sub, self.black_line, self.white_line))
-        self.threshold = [a+b/2 for a,b in zip(self.white_line, self.range_sensor)]
-
     def read_raw_sensors(self):
         """
-        Read the 5 IR sensors of the `Line Follower`_ sensor.
+        Read the 5/6 IR sensors of the `Old Line Follower`_/`Line Follower`_ sensor.
```
Comment (CleoQc) [inline Software/Python/easysensors.py]: red vs black

File: docs/source/api-basic/easygopigo3.rst [hunk 8]
```diff
@@ -26,7 +26,8 @@ EasyGoPiGo3
 .. _grove motion sensor: https://www.dexterindustries.com/shop/grove-pir-motion-sensor/
 .. _grove dht sensor: https://www.dexterindustries.com/shop/temp-humidity/
 .. _servo: https://www.dexterindustries.com/shop/servo-package/
-.. _line follower: https://www.dexterindustries.com/shop/line-follower-for-gopigo/
+.. _line follower: https://www.dexterindustries.com/shop/line-follower-sensor
+.. _old line follower: https://www.dexterindustries.com/shop/line-follower-for-gopigo/
```
Comment (CleoQc) [inline docs/source/api-basic/easygopigo3.rst]: red

File: docs/source/api-basic/sensors.rst [hunk 9]
```diff
@@ -125,7 +125,8 @@ Remote
 .. _grove motion sensor: https://www.dexterindustries.com/shop/grove-pir-motion-sensor/
 .. _grove dht sensor: https://www.dexterindustries.com/shop/temp-humidity/
 .. _servo: https://www.dexterindustries.com/shop/servo-package/
-.. _line follower: https://www.dexterindustries.com/shop/line-follower-for-gopigo/
+.. _line follower: https://www.dexterindustries.com/shop/line-follower-sensor
+.. _old line follower: https://www.dexterindustries.com/shop/line-follower-for-gopigo/
```
Comment (CleoQc) [inline docs/source/api-basic/sensors.rst]: red

File: Software/Python/easysensors.py (hunk 6 above)
Comment (RobertLucian) [reply Software/Python/easysensors.py]: Well, it is - cause you don't have to do anything about it - it's all selected automatically. And `module_id=2` stands for the new/black line follower, which is the only one available on the GoPiGo3. You wouldn't want to let it choose the module automatically when you already know it has to be the new/black line follower. Inherently speaking, letting it do it automatically can be more prone to errors than selecting the module manually.

If you read the documentation for the `module_id` parameter, you'll see it represents 2 modes of operation: automatic and manual. Value `-1` stands for automatic, `1` for the old/red line follower and `2` for the new/black line follower.

File: Projects/PIDLineFollower/README.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
+## PID-based Controller for the Line Follower
+
+This is a project you can use to tune your GoPiGo + (Old) Line Follower to follow a black/white line.
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply Projects/PIDLineFollower/README.md]: Okay, what do you say calling it the "red line follower (the old one)" and "black line follower (the new one)" everywhere? We are already using this naming scheme in other places.

[the remark this answers is not in the document]
Comment (CleoQc) [reply Projects/PIDLineFollower/README.md]: New and old are relative terms.  In 2 years, we will not refer to the black one as being new. Some user out there will take their GPG out of the box for the first time and assume their black one to be the old one . 
You  can refer to them by version number or by color.

**Pull request merged**